x-pack/auditbeat/tracing/perfevent.go

```diff
@@ -0,0 +1,509 @@
+		c.evs = append(c.evs, ev)
+
+		if !doGroup {
+			if err := ev.MapRingNumPages(c.mappedPages); err != nil {
+				return errors.Wrap(err, "perf channel mapring failed")
+			}
+		}
+	}
+	return nil
+}
+
+// C returns the channel to read samples from.
+func (c *PerfChannel) C() <-chan interface{} {
+	return c.sampleC
+}
+
+// ErrC returns the channel to read errors from.
+func (c *PerfChannel) ErrC() <-chan error {
+	return c.errC
+}
+
+// LostC returns the channel to read lost samples notifications.
+func (c *PerfChannel) LostC() <-chan uint64 {
+	return c.lostC
+}
+
+// Run enables the configured probe and starts receiving perf events.
+// sampleC is the channel where decoded perf events are received.
+// errC is the channel where errors are received.
+//
+// The format of the received events depends on the Decoder used.
+func (c *PerfChannel) Run() error {
+	if !atomic.CompareAndSwapUintptr(&c.running, 0, 1) {
+		return ErrAlreadyRunning
+	}
+	c.sampleC = make(chan interface{}, c.sizeSampleC)
+	c.errC = make(chan error, c.sizeErrC)
+	c.lostC = make(chan uint64, c.sizeLostC)
+
+	for _, ev := range c.evs {
+		if err := ev.Enable(); err != nil {
+			return errors.Wrap(err, "perf channel enable failed")
+		}
+	}
+	c.wg.Add(1)
+	go c.channelLoop()
+	return nil
+}
+
+// Close closes the channel.
+func (c *PerfChannel) Close() error {
+	if atomic.CompareAndSwapUintptr(&c.running, 1, 2) {
+		close(c.done)
+		c.wg.Wait()
+		defer close(c.sampleC)
+		defer close(c.errC)
+		defer close(c.lostC)
+	}
+	var errs multierror.Errors
+	for _, ev := range c.evs {
+		if err := ev.Disable(); err != nil {
+			errs = append(errs, err)
+		}
+		if err := ev.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs.Err()
+}
+
+// doneWrapperContext is a custom context.Context that is tailored to
+// perf.Event.ReadRawRecord needs. It's used to avoid an expensive allocation
+// before each call to ReadRawRecord while providing termination when
+// the wrapped channel closes.
+type doneWrapperContext <-chan struct{}
+
+// Deadline always returns no deadline.
+func (ctx doneWrapperContext) Deadline() (deadline time.Time, ok bool) {
+	// No deadline
+	return deadline, false
+}
+
+// Done returns the underlying done channel.
+func (ctx doneWrapperContext) Done() <-chan struct{} {
+	return ctx
+}
+
+// Err returns context.Canceled if the underlying done channel is closed.
+func (ctx doneWrapperContext) Err() error {
+	select {
+	case <-ctx.Done():
+		return context.Canceled
+	default:
+	}
+	return nil
+}
+
+// Value always returns nil.
+func (ctx doneWrapperContext) Value(key interface{}) interface{} {
+	return nil
+}
+
+func makeMetadata(eventID int, record *perf.SampleRecord) Metadata {
+	return Metadata{
+		StreamID:  record.StreamID,
+		Timestamp: record.Time,
+		TID:       record.Tid,
+		PID:       record.Pid,
+		EventID:   eventID,
+	}
+}
+
+func (c *PerfChannel) channelLoop() {
+	defer c.wg.Done()
+	ctx := doneWrapperContext(c.done)
```

Comment:
Would it make sense to make `Run()` receive a `context.Context` that gets propagated till here?

Reply:
Right now the done channel is internal to PerfChannel. The proper way of stopping it is calling its Close() method. I prefer to keep it this way because there is cleanup to do in Close() even when Run() has not been called yet.